src/inmanta/deploy/scheduler.py

```diff
@@ -345,18 +400,43 @@ async def _new_version(
 
     def _start_for_agent(self, agent: str) -> None:
         """Start processing for the given agent"""
-        self._workers[agent] = asyncio.create_task(self._run_for_agent(agent))
+        self._workers[agent] = TaskRunner(endpoint=agent, scheduler=self)
+        self._workers[agent].start()
 
-    async def _run_for_agent(self, agent: str) -> None:
-        """Main loop for one agent"""
-        while self._running:
-            task: Task = await self._work.agent_queues.queue_get(agent)
-            try:
-                await task.execute(self, agent)
-            except Exception:
-                LOGGER.exception("Task %s for agent %s has failed and the exception was not properly handled", task, agent)
+    async def should_be_running(self, endpoint: str) -> bool:
+        """
+        Return if the agent (or the Scheduler if endpoint == Scheduler id) should be running. This will also
+            check if the environment is halted
+
+        :param endpoint: The name of the agent
+        """
+        current_environment = await Environment.get_by_id(self.environment)
+        assert current_environment
+        current_agent = await data.Agent.get(env=self.environment, endpoint=endpoint)
+        if current_agent is None:
+            return False
+        return not (current_environment.halted or current_agent.paused)
+
+    async def refresh_agent_state_from_db(self, name: str) -> None:
+        """
+        Refresh from the DB (authoritative entity) the actual state of the agent.
+            - If the agent is not paused: It will make sure that the agent is running.
+            - If the agent is paused: Stop a particular agent.
+
+        :param name: The name of the agent
+        """
+
+        if name in self._workers:
+            await self._workers[name].notify()
+            return
+
+        requested_agent = await data.Agent.get(env=self.environment, endpoint=name)
+        if requested_agent:
+            self._start_for_agent(agent=requested_agent.name)
```

Comment:
Only if it is not paused in the DB I would assume ?

Reply:
This one was safe because even if a `TaskRunner` is started, it will first check its actual state from the DB, so it would see at some point that it is not allowed to run. But I refactored that part, to avoid making an additional sql request if we already have the information.